hoot-core/src/main/cpp/hoot/core/io/OsmGbdxXmlWriter.cpp

```diff
@@ -0,0 +1,705 @@
+      _writer->writeCDATA(removeInvalidCharacters(val));
+      _writer->writeEndElement();
+      continue;
+    }
+
+    // Image, Platform and Instrument can be a list
+//    if (key == "Src_imgid" || key == "Pltfrm_id" || key == "Ins_Type")
+//    {
+//      QStringList l = val.split(";");
+
+//      _writer->writeStartElement(key);
+//      _writer->writeStartElement("t1");
+//      _writer->writeCharacters(removeInvalidCharacters(l[0]));
+//      _writer->writeEndElement();
+//      _writer->writeStartElement("t2");
+//      // Look for a second value
+//      if (l.size() == 2)
+//      {
+//        _writer->writeCharacters(removeInvalidCharacters(l[1]));
+//      }
+//      else
+//      {
+//        _writer->writeCharacters("NULL");
+//      }
+//      _writer->writeEndElement(); // t2
+//      _writer->writeEndElement(); // key
+//      continue;
+//    }
+
+    // Keywords can be a list
+    if (key == "Kywrd" || key == "Src_imgid" || key == "Pltfrm_id" || key == "Ins_Type")
+    {
+      _writer->writeStartElement(key);
+      QStringList l = val.split(";");
+      for (int i = 0; i < l.size(); ++i)
+      {
+        _writer->writeStartElement("value");
+        _writer->writeCharacters(removeInvalidCharacters(l[i]));
+        _writer->writeEndElement();
+      }
+      _writer->writeEndElement();
+      continue;
+    }
+
+    // Write out the Tag
+    _writer->writeStartElement(removeInvalidCharacters(key));
+    _writer->writeCharacters(removeInvalidCharacters(val));
+    _writer->writeEndElement();
+  } // End tag loop
+
+//  // This is the next to fix.
+//  if (type == ElementType::Relation)
+//  {
+//    ConstRelationPtr relation = boost::dynamic_pointer_cast<const Relation>(element);
+//    if (relation->getType() != "")
+//    {
+//      _writer->writeStartElement("Got Relation");
+//      _writer->writeAttribute("k", "type");
+//      _writer->writeAttribute("v", removeInvalidCharacters(relation->getType()));
+//      _writer->writeEndElement();
+//    }
+//  }
+
+}
+
+void OsmGbdxXmlWriter::_writeNodes(ConstOsmMapPtr map)
+{
+  NoInformationCriterion filter;
+  QList<long> nids;
+  const NodeMap& nodes = map->getNodes();
+  for (NodeMap::const_iterator it = nodes.begin(); it != nodes.end(); ++it)
+  {
+    if (filter.isNotSatisfied(map->getNode(it->first)))
+        nids.append(it->first);
+  }
+
+  // sort the values to give consistent results.
+  qSort(nids.begin(), nids.end(), qLess<long>());
+  for (int i = 0; i < nids.size(); i++)
+  {
+    _newOutputFile();
+    writePartial(map->getNode(nids[i]));
+  }
+}
+
+void OsmGbdxXmlWriter::_writeWays(ConstOsmMapPtr map)
+{
+  const WayMap& ways = map->getWays();
+
+  for (WayMap::const_iterator it = ways.begin(); it != ways.end(); ++it)
+  {
+    ConstWayPtr w = it->second;
+    //  Skip any ways that have parents
+    set<ElementId> parents = map->getParents(w->getElementId());
+    if (parents.size() > 0)
+      continue;
+    if (w.get() == NULL)
+      continue;
+
```

Comment:
The chunk of code above this comment should be reorganized so that the check `w.get() == NULL` happens first. Otherwise, if w *is* null, the program will bomb out with a null pointer exception when w->getElementId() is called. Maybe like this:
```
ConstWayPtr w = it->second;

// Skip if null
if (w.get() == NULL)
  continue;

//  Skip any ways that have parents
set<ElementId> parents = map->getParents(w->getElementId());
if (parents.size() > 0)
  continue;
```

Reply:
Done.